network: make auto_init_ng_netif less board-dependant

kaspar030 opened the pull request:
PR for #2900, IMHO improves #2891.

~~This is work in progress, let's please discuss the concept.~~

I've adapted ng_at86rf2xx and xbee ~~but only the iot-lab_M3 node.~~
Couldn't test xbee for lack of device.

authmillenon commented on boards/iot-lab_M3/include/ng_at86rf2xx_params.h:
```diff
@@ -0,0 +1,22 @@
+/**
+ * @name AT86RF231 configuration
+ */
+    static const  at86rf2xx_params_t at86rf2xx_params[] =
```
indentation

authmillenon commented:
Can we find a solution that has less code duplication AND does not need any arrays stored in memory?

kaspar030 commented:
@authmillenon Where does this store arrays in memory?

Only the device structs are stored in memory, but they have to.

authmillenon commented on boards/iot-lab_M3/include/ng_at86rf2xx_params.h:
```diff
@@ -0,0 +1,22 @@
+            AT86RF231_SLEEP,
+            AT86RF231_SPI_CLK,
+        },
+    };
```
> @authmillenon Where does this store arrays in memory?

here

kaspar030 replied:
This get's optimized out.

authmillenon commented:
> Only the device structs are stored in memory, but they have to.

Yes but not for the whole run-time as global constants IMHO.

kaspar030 commented:
>>  Only the device structs are stored in memory, but they have to.

> Yes but not for the whole run-time as global constants IMHO.

They are driver-specific state variables. If they get stored in memory, that's how the driver has been implemented.

authmillenon commented:
Have look at this example:

```C
#include <stdio.h>

#define ARRAY_SIZE  (2)

struct test {
    int one;
    int two;
};

const struct test t_array[] = {
    { 1, 2 },
    { 2, 3 },
};

void output(const struct test *t, size_t t_size) {
    for (int i = 0; i < t_size; i++) {
        printf("one: %d; two: %d\n", t[i].one, t[i].two);
    }
}

int main(int argc, char **argv) {
    output(t_array, ARRAY_SIZE);
}
```

```
   text	   data	    bss	    dec	    hex	filename
   1406	    560	      8	   1974	    7b6	main
```

vs.

```C
#include <stdio.h>

#define ARRAY_SIZE  (2)

struct test {
    int one;
    int two;
};

#define ARRAY (struct test[]){ \
    { 1, 2 }, \
    { 2, 3 }, \
}

void output(const struct test *t, size_t t_size) {
    for (int i = 0; i < t_size; i++) {
        printf("one: %d; two: %d\n", t[i].one, t[i].two);
    }
}

int main(int argc, char **argv) {
    output(ARRAY, ARRAY_SIZE);
}
```

```
   text	   data	    bss	    dec	    hex	filename
   1394	    560	      8	   1962	    7aa	main
```

authmillenon commented:
Interesting… I'm a little bit surprised, that the text section is shrinking... not the bss section oO

Compiler was `clang` btw

authmillenon commented:
However, the current device init-functions do not allow for that currently. But since you introduce this structs anyways, we can save parameters and just define them with your `param` struct.

authmillenon commented:
You are however propably right about the optimization when I look at the bss section ;-)

kaspar030 commented:
@authmillenon gcc compiles like this: (main.c is the const array example, main2.c the second with ```#define ARRAY```):

<pre>
[kaspar@localhost hello-world]$ gcc -std=c99 main.c -Os -o main
[kaspar@localhost hello-world]$ gcc -std=c99 main2.c -Os -o main2
[kaspar@localhost hello-world]$ size main
   text	   data	    bss	    dec	    hex	filename
   1360	    560	      8	   1928	    788	main
[kaspar@localhost hello-world]$ size main2
   text	   data	    bss	    dec	    hex	filename
   1366	    560	      8	   1934	    78e	main2
[kaspar@localhost hello-world]$ 
</pre>

authmillenon commented on sys/auto_init/auto_init_ng_netif.c:
```diff
@@ -0,0 +1,99 @@
+static void start_nomac(char *stack, int stack_size, ng_netdev_t *device, const char* name) {
+    /* starting NOMAC */
+    DEBUG("Start NOMAC layer for dev %s\n", name);
+    ng_nomac_init(stack, stack_size, MAC_PRIO, name,
```
What if other MAC layers than `nomac` want to be used?

kaspar030 replied:
Uff. All the configurability, all the time.

Either ```#define AT86RF2XX_MAC NOMAC``` or another field in ```*_params```, then a switch(...) over it. The first would be global for all devices of one type, the second would allow for two devices with different MAC layers.

Let's cross that bridge as soon as we have a) a second MAC layer, b) a use case.

OlegHahm replied:
Basically we need a mapping from MAC to device somewhere. I think it's safe to neglect the case of multiple radios of the same type with different MAC protocols for now.

haukepetersen replied:
I think it must be possible for a board, to specify the MAC layer that should be used, e.g. BOARD A, driver A, MAC A - BOARD B, driver A, MAC B... That is something, that we will see in the foreseeable future...

OlegHahm replied:
Sure.

OlegHahm replied:
Even for the same board this is required.

kaspar030 commented:
> However, the current device init-functions do not allow for that currently. But since you introduce this structs anyways, we can save parameters and just define them with your param struct.

well, that could be a next step.

kaspar030 commented:
So you approve the concept? @haukepetersen what do you think?

authmillenon commented:
> So you approve the concept?

Only problem laying still around is https://github.com/RIOT-OS/RIOT/pull/2901#discussion_r29499407

kaspar030 commented:
> Only problem laying still around is #2901 (comment)

Easy to solve as soon as there's a second option for the MAC layer.

authmillenon commented:
I'd prefer to see your solution now :-).

kaspar030 commented:
It's already in there and works with all available MAC layers. )

authmillenon commented:
No seriously? How do you propose to work with multiple MAC layers? Let's say the user wants `at86rf2xx[0]` to use `nomac`, `at86rf2xx[1]` `ieee802154e`, and `xbee[0]` `nomac`, too.

kaspar030 commented:
in e.g., xbee_params.h:
<pre>
static const netlayer_nr_typt_t xbee_mac_layers[XBEE_NUM] = { MAC_LAYER_NOMAC, MAC_LAYER_802154 }
</pre>

in auto_init_ng_netif.c:

<pre>
maclayer_start_func
maclayer_start_functions[xbee_mac_layers[i]](xbee_dev);
</pre>

kaspar030 commented:
But anyhow, I'd rather go for selecting a sane standard MAC layer and not use auto_init for the special cases. Choice is good but not if it complicates 99% of the use cases.

OlegHahm commented on sys/auto_init/auto_init_ng_netif.c:
```diff
@@ -0,0 +1,99 @@
+#include "net/ng_nomac.h"
+#include "net/ng_netbase.h"
+
+#define ENABLE_DEBUG 1
```
`(0)`

OlegHahm commented on sys/auto_init/auto_init_ng_netif.c:
```diff
@@ -0,0 +1,99 @@
+ * @{
+ */
+#define MAC_STACKSIZE           (KERNEL_CONF_STACKSIZE_DEFAULT)
+#define MAC_PRIO                (PRIORITY_MAIN - 3)
```
What's the advantage of defining these values here instead of defining inside the MAC implementation?

authmillenon replied:
At least for nomac it doesn't make sense to make the stack size MAC depend, since most stack usage is generated by calling the device driver (which makes the stack size dependent on the device). The priority is outside of the MAC layer since a user might want to have different priorities for different MAC threads of the same MAC implementation

OlegHahm replied:
> The priority is outside of the MAC layer since a user might want to have different priorities for different MAC threads of the same MAC implementation

Can you give a use case for this?

authmillenon replied:
Border router: Traffic from PAN will probably be bigger that from the Internet (or vice-versa?) => Higher priority for the PAN's `nomac` thread.

OlegHahm replied:
Ok, I didn't assumed to run `nomac` on any device in the long run.

authmillenon replied:
So you want to set a MAC layer on every device, even if it ships with its own MAC layer (like most Ethernet devices e.g. do)?

OlegHahm replied:
No, but I don't want a dummy MAC layer for these devices, either.

authmillenon replied:
But `nomac` isn't a dummy MAC layer (though I don't know exactly what you mean by that). It's an `netapi` ⇔ `netdev` adaption MAC layer.

OlegHahm replied:
Ok, let's keep the definition here.

haukepetersen replied:
as @authmillenon said... But you can of course always implement your device driver by implementing `netapi` directly and skipping the whole `netdev` and MAC part if you feel it fits better...

OlegHahm commented on boards/iot-lab_M3/include/ng_at86rf2xx_params.h:
```diff
@@ -0,0 +1,22 @@
+    };
+/** @} */
+
+#define AT86RF2XX_NUM 1
```
parens.

Kijewski replied:
y?

OlegHahm commented on sys/auto_init/auto_init_ng_netif.c:
```diff
@@ -0,0 +1,99 @@
+/**
+ * @def number of nomac threads we'll initialize
+ */
+#define NOMAC_NUM AT86RF2XX_NUM + XBEE_NUM
```
parens

OlegHahm commented:
> How do you propose to work with multiple MAC layers? Let's say the user wants at86rf2xx[0] to use nomac, at86rf2xx[1] ieee802154e, and xbee[0] nomac, too.

Sounds like pretty theoretical use case to me. I think one could configure this somehow, but I agree with @kaspar030 that we can discuss this as soon as there is a need for it. It doesn't invalidate this concept. Allowing multiple MAC layers (for different transceivers) however, should be possible and I think the proposed solution could do the trick.

kaspar030 commented:
I would like to add MAC layer choice as soon as there's a second MAC layer. The additional code would look pretty sad before when compiled...

kaspar030 commented:
- addressed coments

haukepetersen commented:
I am sorry @kaspar030, but I truly dislike your approach. I my opinion, it is

1. more complicated, in terms as it is way harder to understand and use
2. is missing some important details:
   - MAC layer configuration. And MAC layers will have differing `_init()` function signatures... And just because there is no other MAC layer at the moment does not mean we should not make sure now that we can cope with upcoming ones...
   - does not allow to overwrite the initialization from the application

Especially the last point is very important! The default configuration for a board is fine, but there must be means to (easily) overwrite this to your application needs!

haukepetersen commented:
and by the way, something is wrong, when compiling against the (former) UDP example (#2895), the network device is not initialized...

haukepetersen commented:
Just thinking about the code duplication. Say I change a parameter for the driver init functions, then I would need to change every `xx_param.h` file. Say we introduce something with Parameters for MAC layers in the same fashion, same problem there. So I don't really see the benefit in your approach, the difference is that you duplicate configuration structs and I duplicate a few (1-3 very simple) lines of C code...

haukepetersen commented:
ups, forget the comment about the application specific initialization, I overlooked that part...

kaspar030 commented:
> and by the way, something is wrong, when compiling against the (former) UDP example (#2895), the network device is not initialized...

It's WIP, never tested udp. Will check this.

haukepetersen commented:
No problem. Just tried it for a quick comparison of code sizes between our solutions :-)

OlegHahm commented:
Ok, let's see, what do we need:
1. something that auto-initializes the driver
2. something that auto-initializes the MAC protocol

The first thing is pretty typical task for the board implementation.  The second task should go IMO into `sys/auto_init`. I'm not sure that an application will need to overwrite this. (if it really needs to do this, it could still disable auto_init at all.)

So, what's wrong with configuring somewhere(TM) which radio should use which MAC protocol and run the corresponding init functions in `auto_init`? The sane defaults can go into a header in the board specific part that may be overwritten by another module or the application.

If I look at the current code and pick two arbitrary boards, I see six lines of effective code where only one line differs. I think we can improve this.

haukepetersen commented:
The problem I still see is that we need a flexible approach when it comes to MAC layer and device drivers. For one there will be device drivers that do implement the `netapi` interface directly, and hence do not need (nor work) with a MAC layer. Also I see use cases, where we have different interfaces on a board that use different MAC layers (here we see what the first devices with BLE and 802.15.4 will bring).

> If I look at the current code and pick two arbitrary boards, I see six lines of effective code where only one line differs. I think we can improve this.

This is true, but you look at two very similar boards/radios that are initialized... The init code for say the mbed board with ethernet and xbee will probably look slightly different :-)

So all I am saying is that we need a very flexible approach that is easy to use and handle.

kaspar030 commented:
> Just thinking about the code duplication. Say I change a parameter for the driver init functions, then I would need to change every xx_param.h file. 

In your approach you'd have to change every auto_init_board.c *and* every auto_init_application.c file.

If the set of minimal parameters per device changes, there is NO way around changing that set of parameters everywhere it is supposed to be set. The reason why you have to touch a .h file, the SPI defines are in ```board.h```, and not ```periph_conf.h```, are somewhere else.

I agree the structs have to move to "periph_conf.h", and replace the board specific defines. But for that some refactoring is needed.

> Say we introduce something with Parameters for MAC layers in the same fashion, same problem there. So I don't really see the benefit in your approach, the difference is that you duplicate configuration structs and I duplicate a few (1-3 very simple) lines of C code...

Those 1-3 very simple lines of code are *all* that is necessary to skip auto config and do it by hand. It is not more. If the default doesn't cut it, initialize yourself.
*but* my approach already supports N devices per driver. To add that to your ( 1-3 very simple) lines of code, you end up with 2-5 at least. All that gets multiplied with the number of boards. Duplicated *logic*.

Also already in your 1-3 simple lines of code you introduce 2k MAC stack for one board, 1k MAC stack in the driver test override. Was that on purpose? If yes, it is impossible to spot. If not, it more than proves my point.

> 1. more complicated, in terms as it is way harder to understand and use

I don't buy that. You just write out the complete configuration N times, in C code. You would end up with NUM_BOARD + NUM_APP network setup routines.
My approach has a struct per device, and all devices get auto-setup with a default MAC layer. That means, there is exactly one network setup routine in the repo.

> is missing some important details: MAC layer configuration. And MAC layers will have differing _init() function signatures... And just because there is no other MAC layer at the moment does not mean we should not make sure now that we can cope with upcoming ones...

How do you cope with this when having a .c file per board?

kaspar030 commented:
> The problem I still see is that we need a flexible approach when it comes to MAC layer and device drivers. For one there will be device drivers that do implement the netapi interface directly, and hence do not need (nor work) with a MAC layer. Also I see use cases, where we have different interfaces on a board that use different MAC layers (here we see what the first devices with BLE and 802.15.4 will bring).

I agree. But this is a use case that will never be visible in your .c files. You'd always have to rewrite your "1-3 simple lines of code" to reflect it. Same as with my aproach.

For this we should have an example application that doesn't use the auto-init module, but shows how simple it is to do by hand.

All I did was put the default configuration in a central place.

kaspar030 commented:
> No problem. Just tried it for a quick comparison of code sizes between our solutions :-)

Works fine here. Did you add "auto_init_ng_netif" to USEMODULES?

haukepetersen commented:
> In your approach you'd have to change every auto_init_board.c and every auto_init_application.c file.

Not quite, it is 'every auto_init_board.c and every auto_init_application.c file' that contains calls to the driver/MAC layer that was changed (a quite significant difference).

> If the set of minimal parameters per device changes, there is NO way around changing that set of parameters everywhere it is supposed to be set. The reason why you have to touch a .h file, the SPI defines are in board.h, and not periph_conf.h, are somewhere else.

I do not quite get what you are saying here, especially the second sentence. In my version you have to touch some C files, in your version you have to touch headers, or how is that meant?

> I agree the structs have to move to "periph_conf.h", and replace the board specific defines. But for that some refactoring is needed.

Noo, this stuff is not meant to go into the `periph_conf.h`. That file includes the PERIPHERAL configuration, and device drivers are not part of that...

> Those 1-3 very simple lines of code are all that is necessary to skip auto config and do it by hand. It is not more. If the default doesn't cut it, initialize yourself.
but my approach already supports N devices per driver. To add that to your ( 1-3 very simple) lines of code, you end up with 2-5 at least. All that gets multiplied with the number of boards. Duplicated logic.

But same in your solution? You have to define the param_xx.h header for every board, too.

> Also already in your 1-3 simple lines of code you introduce 2k MAC stack for one board, 1k MAC stack in the driver test override. Was that on purpose? If yes, it is impossible to spot. If not, it more than proves my point.

The other way around, because the test application was used to debug the driver and therefore I put the stacksize to printf. Easy to spot :-)

> I don't buy that. You just write out the complete configuration N times, in C code. You would end up with NUM_BOARD + NUM_APP network setup routines.
My approach has a struct per device, and all devices get auto-setup with a default MAC layer. That means, there is exactly one network setup routine in the repo.

Wrong, not 'NUM_BOARD + NUM_APP', but '# of boards with network interfaces + some applications that do their own stuff'. This is exactly the same amount of header `param` header files one has to write with your approach... Except that in my approach you are completely flexible in doing whatever you need instead of using pre-defined options that you can only change by using some defines that are more difficult to understand...

> How do you cope with this when having a .c file per board?

You just change your board to whatever the new default MAC layer is. Actually I expect NOMAC to stay the default MAC for many boards, so you will probably never have to change all the .c files...

haukepetersen commented:
> I agree. But this is a use case that will never be visible in your .c files. You'd always have to rewrite your "1-3 simple lines of code" to reflect it. Same as with my aproach.

The use case is not to change this in you application, but to have this as default application: use as default device A with MAC X and device B with MAC Y -> set up once and let it be.

> All I did was put the default configuration in a central place

Not really, you put the init-code into a central place and defined a set of device-specific configuration values - IMHO taking away ease and flexibility.

haukepetersen commented:
Yap, I forgot the USEMODULE...

yours:
```
   text	   data	    bss	    dec	    hex
  44124	    164	  17764	  62052	   f264
```
mine:
```
   text	   data	    bss	    dec	    hex
  44116	    164	  17764	  62044	   f25c	
```
(on iot-lab_M3 using gcc)

haukepetersen commented:
Oh, one more thing (but that could be just an impression): The driver initialization file reminds me heavily on the `sys/transceiver/transceiver.c`. Or at least it seems to me that the file is very prone to move in that direction once multiple people start to add their stuff to it...

haukepetersen commented:
Anyway, I seem to sense that we don't get to the same opinion on this anytime soon. Maybe is there an altogether different way to do this stuff we are both happy with?

kaspar030 commented:
> The reason why you have to touch a .h file, the SPI defines are in board.h, and not periph_conf.h, are somewhere else.
> 
> I do not quite get what you are saying here, especially the second sentence. In my version you have to touch some C files, in your version you have to touch headers, or how is that meant?

The header struct holds the device specific parameters. They *should* replace ```#define AT86RF231_SPI SPI_0```. If a board does have

kaspar030 commented:
>> Those 1-3 very simple lines of code are all that is necessary to skip auto config and do it by hand. It is not more. If the default doesn't cut it, initialize yourself.
>>     but my approach already supports N devices per driver. To add that to your ( 1-3 very simple) lines of code, you end up with 2-5 at least. All that gets multiplied with the number of boards. Duplicated logic.

> But same in your solution? You have to define the param_xx.h header for every board, too.

The difference is that one is static, per-board data, the other is logic + data with only the data imprinted in the logic is board-specific.

If you take your code and factor out the parts that are actually board-specific, you end up with the static data in my approach.

kaspar030 commented:
>> If the set of minimal parameters per device changes, there is NO way around changing that set of parameters everywhere it is supposed to be set. The reason why you have to touch a .h file, the SPI defines are in board.h, and not periph_conf.h, are somewhere else.

> I do not quite get what you are saying here, especially the second sentence. In my version you have to touch some C files, in your version you have to touch headers, or how is that meant?

The point is that if the driver needs a different set of parameters and that set is board specific, some file in that board's folder has to be changed. We will not find a solution without that. This is the reason why there is board-specific .c/.h stuff.

Nothing in board/ should have influence on the priority of a MAC layer, though. That has to be abstracted away and put in a central place.

kaspar030 commented:
> Yap, I forgot the USEMODULE...
> 
> yours:
> 
>    text    data     bss     dec     hex
>   44124     164   17764   62052    f264
> 
> mine:
> 
>    text    data     bss     dec     hex
>   44116     164   17764   62044    f25c 
> 
> (on iot-lab_M3 using gcc)

Did you disable the version string in kernel_init.c?

kaspar030 commented:
> Wrong, not 'NUM_BOARD + NUM_APP', but '# of boards with network interfaces + some applications that do their own stuff'. This is exactly the same amount of header param header files one has to write with your approach... Except that in my approach you are completely flexible in doing whatever you need instead of using pre-defined options that you can only change by using some defines that are more difficult to understand...

That header defines the pin mapping of a device. You might argue that you don't like typechecked, machine-iterable ways of defining static data and would prefer to use #defines. That is not the point here. Somewhere in the board/ config we will save that board's network device's pin mapping.

And again - just because you already copied the code to every location where you might need flexibility doesn't mean that that is more flexible than my approach. The only real difference is that I tried to remove the duplicated code, so it scales to the hundreds of boards we might support in the future.
It is *always* possible to not use auto_init and use *exactly* your code to initialize a board's network devices.

kaspar030 commented:
>>  How do you cope with this when having a .c file per board?

> You just change your board to whatever the new default MAC layer is. Actually I expect NOMAC to stay the default MAC for many boards, so you will probably never have to change all the .c files...

Seriously, I prefer one C file we don't have to change than 100.

kaspar030 commented:
> Anyway, I seem to sense that we don't get to the same opinion on this anytime soon. Maybe is there an altogether different way to do this stuff we are both happy with?

Doesn't work this way...

> Ok, let's see, what do we need:
> 1. something that auto-initializes the driver
> 2. something that auto-initializes the MAC protocol

Both approaches can do both. Both have pretty much the same code size and use exactly the same amount of RAM. Both completely allow overriding.

The difference is the code duplication. That's why I started.

kaspar030 commented:
(sorry, closed by accident)

kaspar030 commented:
> Wrong, not 'NUM_BOARD + NUM_APP', but '# of boards with network interfaces + some applications that do their own stuff'. This is exactly the same amount of header param header files one has to write with your approach...

It's not the number of files, it's the content.

kaspar030 commented:
>>    I agree. But this is a use case that will never be visible in your .c files. You'd always have to rewrite your
> "1-3 simple lines of code" to reflect it. Same as with my aproach.

> The use case is not to change this in you application, but to have this as default application: use as default device A with MAC X and device B with MAC Y -> set up once and let it be.

How do you set it up with your approach without writing the whole board-specific setup?

kaspar030 commented:
> Oh, one more thing (but that could be just an impression): The driver initialization file reminds me heavily on the sys/transceiver/transceiver.c. Or at least it seems to me that the file is very prone to move in that direction once multiple people start to add their stuff to it...

The auto_initialization can be spread in per-driver files if necessary.

OlegHahm commented:
> > Ok, let's see, what do we need:
    1. something that auto-initializes the driver
    2. something that auto-initializes the MAC protocol

> Both approaches can do both. Both have pretty much the same code size and use exactly the same amount of RAM. Both completely allow overriding.

Technically yes, conceptionally I think it's kind of unclean to initiate a MAC layer in the board dependent code. After thinking about it for a while I think the solution in this PR is much cleaner. However, I agree that we should split the auto_init code into multiple files. The ´ifdef` approach here (and in transceiver) is ugly, but IMO necessary for a clean solution. (Sounds weird, I know: I mean it **looks** ugly but **is** clean.)

kaspar030 commented:
> The ´ifdef` approach here (and in transceiver) is ugly, but IMO necessary for a clean solution.

We could adopt the libc constructor method, which involves writing the device setup functions into a specific linker section, then iterating those at bootup.

But I discussed that with Hauke, we both agree this would be too much magic.

OlegHahm commented:
Yeah, avoid (linkerscript) magic. I think we all made bad experience with stuff like this in the past.

(Actually, I think this should go into the coding conventions.)

haukepetersen commented:
Ok fine, I actually give up and if you want, go with this approach. Although I promise that it will get ugly in the future...

Then lets look at the code and make at least the best of it.

haukepetersen commented on boards/iot-lab_M3/include/ng_at86rf2xx_params.h:
```diff
@@ -0,0 +1,46 @@
+    };
+/** @} */
+
+#define AT86RF2XX_NUM (sizeof(at86rf2xx_params)/sizeof(at86rf2xx_params[0]))
```
this should be moved to `sys/auto_init/auto_init_ng_netif.c`

haukepetersen commented on boards/samr21-xpro/include/ng_at86rf2xx_params.h:
```diff
@@ -0,0 +1,46 @@
+    };
+/** @} */
+
+#define AT86RF2XX_NUM (sizeof(at86rf2xx_params)/sizeof(at86rf2xx_params[0]))
```
move to `sys/auto_init/auto_init_ng_netif.c`

haukepetersen commented on boards/iot-lab_M3/include/ng_at86rf2xx_params.h:
```diff
@@ -0,0 +1,46 @@
+
+/**
+ * @ingroup   board_iot-lab_M3
+ * @brief     at86rf231 board specific configuration
```
`brief` should go after `file`. And some newlines to fit the formatting of all the other RIOT headers would be nice.

haukepetersen commented on boards/samr21-xpro/include/ng_at86rf2xx_params.h:
```diff
@@ -0,0 +1,46 @@
+
+/**
+ * @ingroup   board_samr21-xpro
+ * @brief     at86rf233 board specific configuration
```
same here.

haukepetersen commented on boards/samr21-xpro/include/ng_at86rf2xx_params.h:
```diff
@@ -0,0 +1,46 @@
+extern "C" {
+#endif
+
+#ifndef NG_AT86RF2XX_PARAMS_H
```
this is normally positioned above the cplusplus stuff in RIOT files...

haukepetersen commented on boards/samr21-xpro/include/ng_at86rf2xx_params.h:
```diff
@@ -0,0 +1,46 @@
+#ifdef __cplusplus
+}
+#endif
+/** @} */
```
move this to the end of the file

haukepetersen commented on sys/auto_init/auto_init.c:
```diff
@@ -307,6 +307,7 @@ void auto_init(void)
 #endif
 #ifdef MODULE_AUTO_INIT_NG_NETIF
     DEBUG("Auto init network interfaces.\n");
+    extern void auto_init_ng_netif(void);
```
you can drop this, as the function is defined in `auto_init.h`

haukepetersen commented on sys/auto_init/auto_init_ng_netif.c:
```diff
@@ -0,0 +1,115 @@
+ * @file    auto_init_ng_netif_c
+ * @brief   initializes netif based network interfaces
+ * @author  Kaspar Schleiser <kaspar@schleiser.de>
+ */
```
This header is completely broken, please adapt!

kaspar030 replied:
Yeah, I copied that from auto_init.c ... Does doxygen actually parse .c files?

OlegHahm replied:
No. At least it's configured not to do so.

haukepetersen commented on sys/auto_init/auto_init_ng_netif.c:
```diff
@@ -0,0 +1,115 @@
+/**
+ * @def number of nomac threads we'll initialize
+ */
+#define NOMAC_NUM (AT86RF2XX_NUM + XBEE_NUM)
```
This is gonna look nice once we have 20 radio drivers... :-)

haukepetersen commented on tests/driver_at86rf2xx/ng_at86rf2xx_params.h:
```diff
@@ -0,0 +1,48 @@
+            .cs_pin = GPIO_0,
+            .int_pin = GPIO_1,
+            .sleep_pin = GPIO_2,
+            .reset_pin = GPIO_3,
```
here you break the test for the iot-lab and the samr21! The use different configurations for the device. Also this must be settable in the Makefile, as this driver should also work for boards that do not have the radio on-board, but where the radio is connected via jumper wires to any arbitrary setup.

kaspar030 replied:
No I don't break the tests:
(from the Makefile)
<pre>
ifneq (,$(DRIVER))
  USEMODULE += $(DRIVER)
else
  # default to ng_at86rf231
  USEMODULE += ng_at86rf231

  # This adds . to include path so generic ng_at86rf2xx_params.h gets picked up.
  # All boards actually having such a device on board should
  # define DRIVER above to skip this step, as the board provides this header.
  CFLAGS += -I$(CURDIR)
endif

</pre>

But I do think a variable like "USE_BOARD_DRIVER_PARAMS" would be a lot nicer.

haukepetersen replied:
You don't break it directly, but you change the functionality slightly: If I want to test the radio with say the stm32f4discovery, and I connect it to SPI_1, GPIO_9-12. Then I would now need to change the header file. In the old test, all I had to do is define 5 environment variables on my shell and didn't have to change one line of code/Makefile to run it...

kaspar030 replied:
You're right. Using defines in the header would solve this, right?

haukepetersen replied:
yap

haukepetersen commented on tests/driver_xbee/xbee_params.h:
```diff
@@ -0,0 +1,38 @@
+#endif
+
+static xbee_params_t xbee_params[] = {
+    { .uart = XBEE_UART,
```
same here.

kaspar030 replied:
No, as the Makefile defines XBEE_UART.

kaspar030 commented:
- addressed @haukepetersen's comments
- split up initialization in per-device files

OlegHahm commented on sys/auto_init/netif/auto_init_ng_at86rf2xx.c:
```diff
@@ -0,0 +1,71 @@
+ * @brief   Define stack parameters for the MAC layer thread
+ * @{
+ */
+#define AT86RF2XX_MAC_STACKSIZE     (KERNEL_CONF_STACKSIZE_DEFAULT)
```
I still think this should go into the MAC implementation.

kaspar030 replied:
Well, the MAC implementation calls a driver's functions for e.g., send, receive, etc. So this is low-level driver dependent and will be different for different devices. That's why it is defined here. If defined by nomac, it would have to be a size that fits all drivers, making possible optimization impossible.

OlegHahm replied:
Okay, agreed.

haukepetersen replied:
Actually, the ideal solution would be to compute this number by something like BASE_STACK + DRIVERS_STACK_NEEDS + MAC_LAYER_STACK_NEEDS, where the driver part is defined by the driver and the MAC part is defined by... you get the picture. The question is just if it is possible to define these numbers in a sensible way?

kaspar030 replied:
Hm... One day we need to find a procedure for finding stack sizes for all the little helper threads...

OlegHahm commented:
Btw. is this still WIP (as the label says)?

kaspar030 commented:
@OlegHahm Lets say, no, please review. ;)

OlegHahm commented on sys/auto_init/Makefile:
```diff
@@ -8,4 +8,10 @@ endif
 
 DIRS += $(AUTO_INIT_MODULES)
 
+SRC = auto_init.c
```
What's this for?

kaspar030 replied:
Ah, this is not needed anymore.

OlegHahm commented on sys/auto_init/netif/auto_init_ng_at86rf2xx.c:
```diff
@@ -0,0 +1,71 @@
+
+void auto_init_ng_at86rf2xx(void)
+{
+    int nomac_stacks_used = 0;
```
Actually I think this is superfluous.

kaspar030 replied:
Because it could just use 'i' from the for loop?

OlegHahm replied:
Yes, the memory is "wasted" anyway.

OlegHahm commented:
ACK, Squash and go if Travis is also fine.

kaspar030 commented:
-squashed, rebased

OlegHahm commented:
The `driver_xbee` test fails.

authmillenon commented:
`ng_at86rf2xx_params.h` does not defined for the boards where this test fails.

authmillenon commented:
See https://github.com/kaspar030/RIOT/pull/1

kaspar030 commented:
Should I squash this?

kaspar030 commented:
@OlegHahm did you retry the xbee test?

OlegHahm commented:
I just did. Builds without problem on my system. Pleas squash.

kaspar030 commented:
- squashed

kaspar030 commented:
@haukepetersen *ping*

OlegHahm commented:
Please rebase and add `USEMODULE += fix_auto_init_netif` to `tests/driver_kw2xrf/Makefile`.

ACK then. Let's get this merged.

jremmert-phytec-iot commented:
I already prepared a new PR to adapt ```kw2xrf``` to this auto-init poperly. Will open when this is merged.

OlegHahm commented:
Cool! :+1:

kaspar030 commented:
> add USEMODULE += fix_auto_init_netif to tests/driver_kw2xrf/Makefile.

I fixed the problem differently: the auto_init_* functions where called whenever a radio module was selected, but the auto_init_ng_module wasn't linked in. Fixed. Pls have a quick look, then I'll squash.

OlegHahm commented:
Had a quick look. I'm fine. Go squashing.

kaspar030 commented:
The last commit doesn't squash easily, so I just fixed the commit message.

jremmert-phytec-iot commented:
@kaspar030 @OlegHahm I think Travis had an unrelated error. Could someone kick Travis?

kaspar030 commented:
So this was ACKed? travis is happy...

OlegHahm commented:
I say ACK and as far as I understood @haukepetersen, he doesn't NACK.

kaspar030 commented:
allright, then GO.

OlegHahm commented:
:beers: :balloon: :confetti_ball:
